src/Microsoft.Azure.EventHubs/EventHubsDiagnosticSource.cs

```diff
@@ -0,0 +1,125 @@
+    /// </summary>
+    internal static class EventHubsDiagnosticSource
+    {
+        public const string DiagnosticSourceName = "Microsoft.Azure.EventHubs";
+
+        public const string SendActivityName = DiagnosticSourceName + ".Send";
+        public const string SendActivityStartName = SendActivityName + ".Start";
+        public const string SendActivityExceptionName = SendActivityName + ".Exception";
+
+        public const string ReceiveActivityName = DiagnosticSourceName + ".Receive";
+        public const string ReceiveActivityStartName = ReceiveActivityName + ".Start";
+        public const string ReceiveActivityExceptionName = ReceiveActivityName + ".Exception";
+
+        internal static readonly DiagnosticListener DiagnosticListener = new DiagnosticListener(DiagnosticSourceName);
+
+        internal static bool IsEnabled => DiagnosticListener.IsEnabled();
+
+        internal static Activity StartSendActivity(string clientId, EventHubsConnectionStringBuilder csb, string partitionKey, IEnumerable<EventData> eventDatas, int count)
+        {
+            // skip if diagnostic source not enabled
+            if (!IsEnabled)
+            {
+                return null;
+            }
+
+            // skip if no listeners for this "Send" activity 
+            if (!DiagnosticListener.IsEnabled(SendActivityName,
+                new
+                {
+                    Endpoint = csb.Endpoint,
+                    EntityPath = csb.EntityPath,
+                    PartitionKey = partitionKey
+                }))
+            {
+                return null;
+            }
+
+            Activity activity = new Activity(SendActivityName);
+
+            // extract activity tags from input
+            activity.AddTag("component", "Microsoft.Azure.EventHubs");
+            activity.AddTag("span.kind", "producer");
+            activity.AddTag("operation.name", $"Send");
+            activity.AddTag("operation.data", $"{csb.EntityPath}/{partitionKey}");
+            activity.AddTag("peer.service", "Azure Event Hub");
+            activity.AddTag("peer.hostname", csb.Endpoint.OriginalString);
+            activity.AddTag("eh.event_hub_name", csb.EntityPath);
+            activity.AddTag("eh.partition_key", partitionKey);
+            activity.AddTag("eh.event_count", count.ToString());
+            activity.AddTag("eh.client_id", clientId);
+
+            // in many cases activity start event is not interesting, 
+            // in that case start activity without firing event
+            if (DiagnosticListener.IsEnabled(SendActivityStartName))
+            {
+                DiagnosticListener.StartActivity(activity,
+                    new
+                    {
+                        Endpoint = csb.Endpoint,
+                        EntityPath = csb.EntityPath,
+                        PartitionKey = partitionKey,
+                        EventDatas = eventDatas
```

Comment:
> EventDatas = eventDatas [](start = 23, length = 24)

What are we actually tracing with actual messages? The tracing payload will be huge for all messages.

Reply:
The telemetry listener that we are providing will not look into these, only tags. However DiagnosticSource instrumentation should be generic so that it can be used for different purposes - e.g. on-demand inspection of events being sent or metadata injection. 

Similarly in [System.Net.Http](https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs) the whole ```HttpRequestMessage``` is being passed